ddl/failtest/fail_db_test.go

```diff
@@ -165,6 +165,57 @@ func (s *testFailDBSuite) TestHalfwayCancelOperations(c *C) {
 	_, err = s.se.Execute(context.Background(), "select * from tx")
 	c.Assert(err, IsNil)
 
+	// test for exchanging partition
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/ddl/exchangePartitionErr", `return(true)`), IsNil)
+	defer func() {
+		c.Assert(failpoint.Disable("github.com/pingcap/tidb/ddl/exchangePartitionErr"), IsNil)
+	}()
+	_, err = s.se.Execute(context.Background(), "create table pt(a int) partition by hash (a) partitions 2")
+	c.Assert(err, IsNil)
+	_, err = s.se.Execute(context.Background(), "insert into pt values(1), (3), (5)")
+	c.Assert(err, IsNil)
+	_, err = s.se.Execute(context.Background(), "create table nt(a int)")
+	c.Assert(err, IsNil)
+	_, err = s.se.Execute(context.Background(), "insert into nt values(7)")
+	c.Assert(err, IsNil)
+	_, err = s.se.Execute(context.Background(), "alter table pt exchange partition p1 with table nt")
+	c.Assert(err, NotNil)
+
+	rs, err = s.se.Execute(context.Background(), "select count(*) from pt")
+	c.Assert(err, IsNil)
+	req = rs[0].NewChunk()
+	err = rs[0].Next(context.Background(), req)
+	c.Assert(err, IsNil)
+	c.Assert(req.NumRows() == 0, IsFalse)
+	row = req.GetRow(0)
+	c.Assert(row.Len(), Equals, 1)
+	c.Assert(row.GetInt64(0), DeepEquals, int64(3))
```

Comment:
Why not test all rows? Here is a convenient (hope it work) function for tests: 
```go
session.ResultSetToStringSlice(ctx context.Context, s Session, rs sqlexec.RecordSet) ([][]string, error)
```

Reply:
I have rewritten this function. PTAL @tangenta